runtime-modules/content/src/nft/types.rs

```diff
@@ -169,391 +68,160 @@ impl<MemberId> Default for NftOwner<MemberId> {
     }
 }
 
-/// Information on the auction being created.
-#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
-#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
-pub struct Bid<MemberId, AccountId, BlockNumber: BaseArithmetic + Copy, Balance> {
-    pub bidder: MemberId,
-    pub bidder_account_id: AccountId,
-    pub amount: Balance,
-    pub made_at_block: BlockNumber,
-}
-
-impl<MemberId, AccountId, BlockNumber: BaseArithmetic + Copy, Balance>
-    Bid<MemberId, AccountId, BlockNumber, Balance>
-{
-    fn new(
-        bidder: MemberId,
-        bidder_account_id: AccountId,
-        amount: Balance,
-        made_at_block: BlockNumber,
-    ) -> Self {
-        Self {
-            bidder,
-            bidder_account_id,
-            amount,
-            made_at_block,
-        }
-    }
-}
-
-/// Information on the auction being created.
+/// Parameters, needed for auction start
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 #[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
-pub struct AuctionRecord<
-    BlockNumber: BaseArithmetic + Copy,
-    Balance: Clone,
-    MemberId: Ord + Clone,
-    AccountId: Ord + Clone,
-> {
+pub struct AuctionParamsRecord<AuctionType, Balance, MemberId: Ord> {
+    // Auction type (either english or open)
+    pub auction_type: AuctionType,
```

Comment:
I don't love this approach too be honest. Its now become much harder to cleanly figure out what the type hierarchy actually looks like, and all of the naming is not longer making that much sense. I have spent >20mins now just trying to track things down and understanding all the different meanings of  `auction_type`and `AuctionRecord` vs. `AuctionParamRecord`, which by the way are _literally_ identical!

It seems wrong to me that you are basically introducing a trait parameter which _will only ever be instantiated by one type_. This smells of incorrect usage of the abstractions, the other trait parameters here, like `Balance` or `MemberId`, they could in principle be lots of wacky types when this pallet is used, but `AuctionType` can only ever be one type in any usage of this code.

I am not exactly sure what you are trying to to achieve from your description, but I don't think this is the right way, can you please change it to the direct approach. If you are adamant about there being some serious problem with that approach, then let us do a call.

Reply:
I removed the duplicated type in 4f26837b202f027efed81d26adc45295a8f5ee62
I see the problem you are pointing at. The only draw back I see in changing back to the original type is the following:
``` rust
/// Auction Record alias
pub type Auction<T> =
    AuctionRecord<CurrencyOf<T>, <T as common::MembershipTypes>::MemberId, AuctionTypeOf<T>>;

/// AuctionParameters used for Auction initialization, it is almost a copycat of Auction<T>
/// Except that AuctionTypeOf<T> has been replaced because auction_id member is unknown @ creation
pub type AuctionParams<T> = AuctionRecord<
    CurrencyOf<T>,
    <T as common::MembershipTypes>::MemberId,
    AuctionType<EnglishAuction<T>, <T as frame_system::Trait>::BlockNumber>,
>;
```
While this usage might seem horrendous it does the job, because `auction_id` type inside the `OpenAuction<T>` record cannot be known before creation. And it is needed in order to differentiate between:
 cancelling bid from an ongoing open auction vs. cancelling bid from a past expired open auction.

Reply:
I did not understand this post, what is the draw back of changing back? We are already all over the runtime distingushing between a type and the a type to create that type, auctions are not special.

Reply:
The main problem I have in splitting the 2 auction is situations like the following. I will have to duplicate functions for english / opens auction such as `can_cancel_auction`:
``` rust
            // Ensure auction for given video id exists
            let auction = Self::ensure_auction_state(&nft)?;

            // Ensure given auction can be canceled
            Self::ensure_auction_can_be_canceled(&auction, video_id)?;
```
This does not seems to me sustainable, for example if tomorrow we want to add an extra type of auction this will force to add extra function: `can_ancel_new_auction_variant`

Reply:
No `AuctionType` is not really a sensible type as is. 

I would suggest you do one of two things,
a) totally split up extrincs and types for the two auctions
b) introduce new enum types for auction state and auction creation parameters, across the two parameters. I am not sure this is really worth it, but since you are thinking more in these terms, this seems like a way to do it which does not have the same problems as the prior approach

```Rust

pub enum TransactionalStatusRecord<MemberId, Balance, AuctionType> {
    Idle,
    InitiatedOfferToMember(MemberId, Option<Balance>),
    Auction(Auction),
    BuyNow(Balance),
}

pub enum Auction {
  EnglishAuction(EnglishAuction),
  OpenAuction(OpenAuction)
}

pub struct EnglishAuction {
  /// state of an english auction
} 

pub struct OpenAuction {
 /// state of an open auction
}

pub enum AuctionParams<BlockNumber, Balance, MemberId: Ord> {
  EnglishAuctionParams(EnglishAuctionParams),
  OpenAuctionParams(OpenAuctionParams)
}

pub struct EnglishAuctionParams {
  //// creation info
}

pub struct OpenAuctionParams {
   //// creation info
}
```

there will be some common fields in play here, and if you don't like that you can solve it with some static composition, like

```
pub struct BaseAuctionProperties {
    pub starting_price: Balance,
    pub buy_now_price: Option<Balance>,
    pub whitelist: BTreeSet<MemberId>,
}

pub struct EnglishAuction {
 base_properties: BaseAuctionProperties
 ...
 }
```

but not sure how much value it really has. 

I would say is that the main problem with this entire approach is that its trying to hang on to the idea that open and english auctions are virtually indistinguishable, when that is basically no longer the case.